### PR TITLE
tcp_proxy: fix allowing multiple readDisable calls

### LINKED ISSUE #27516: suspected crash bug: hitting debug assert in tcp_proxy
I set up Envoy with a single listener that has a tcp_proxy filter, with TCP tunneling enabled, where the upstream cluster has ``http2_protocol_options`` set, meaning that upsream codec is expected to be HTTP/2.

The problem I'm facing is that when there's a high throughput load tunneled through the proxy filter, the following line causes the process to crash in debug build:
https://github.com/envoyproxy/envoy/blob/0ea68d2eb56a59be4216fcf57eefe243d5968e19/source/common/tcp_proxy/tcp_proxy.cc#L287

This was also mentioned in #24839, but I think that now after some digging and learning the code, I have better insights, therefore opened a new issue to escalate this. After reading the [flow_control#http2-implementation-details](https://github.com/envoyproxy/envoy/blob/main/source/docs/flow_control.md#http2-implementation-details) to try and understand how the flow control works, I noticed that for HTTP/2 streams it's possible that there are two flow control blocks - one at the network level and one at the H2 level. Considering that if I comment out the assert calls - the tunneling flow works correctly, I think that this is what's happening in this case (will be explained through logs below), and the fix to the code would be either removing the assertions, or changing the assertion so multiple watermark calls may be called.

The following trace logs are provided for the when the assert calls are not commented out (later I commented out the assert calls):

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

```
... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

connection_impl.cc:530       [C2] onAboveReadBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=1048585
tcp_proxy.cc:673             [C2] downstream connection received 1048585 bytes, end_stream=false
connection_impl.cc:523       [C2] onBelowReadBufferLowWatermark
connection_impl.cc:362       [C2] readDisable: disable=false disable_count=1 state=0 buffer_length=0
codec_impl.cc:586            [C3] send buffer over limit 
upstream.cc:119              HttpUpstream::onAboveWriteBufferHighWatermark
tcp_proxy.cc:338             [C2] Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=0

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

codec_impl.cc:593            [C3] send buffer under limit 
upstream.cc:124              HttpUpstream::onBelowWriteBufferLowWatermark
tcp_proxy.cc:349             [C2] Filter::UpstreamCallbacks::onBelowWriteBufferLowWatermark
connection_impl.cc:362       [C2] readDisable: disable=false disable_count=1 state=0 buffer_length=0
connection_impl.cc:483       [C3] writing 16393 bytes, end_stream false
connection_impl.cc:548       [C3] onAboveWriteBufferHighWatermark
upstream.cc:119              HttpUpstream::onAboveWriteBufferHighWatermark
tcp_proxy.cc:338             [C2] Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=0

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

connection_impl.cc:537       [C3] onBelowWriteBufferLowWatermark
upstream.cc:124              HttpUpstream::onBelowWriteBufferLowWatermark
tcp_proxy.cc:349             [C2] Filter::UpstreamCallbacks::onBelowWriteBufferLowWatermark
connection_impl.cc:362       [C2] readDisable: disable=false disable_count=1 state=0 buffer_length=0

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

connection_impl.cc:530       [C2] onAboveReadBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=1048585
tcp_proxy.cc:673             [C2] downstream connection received 1048585 bytes, end_stream=false
connection_impl.cc:523       [C2] onBelowReadBufferLowWatermark
connection_impl.cc:362       [C2] readDisable: disable=false disable_count=1 state=0 buffer_length=0
codec_impl.cc:586            [C3] send buffer over limit             ### Here the high watermark is called by codec_impl
upstream.cc:119              HttpUpstream::onAboveWriteBufferHighWatermark
tcp_proxy.cc:338             [C2] Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=0

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

connection_impl.cc:548       [C3] onAboveWriteBufferHighWatermark    ### Here the high watermark is called by connection_impl
upstream.cc:119              HttpUpstream::onAboveWriteBufferHighWatermark
tcp_proxy.cc:338             [C2] Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark
tcp_proxy.cc:339             assert failure: !on_high_watermark_called_.

backtrace.h:104 Caught Aborted, suspect faulting address 0x19
backtrace.h:91 Backtrace (use tools/stack_decode.py to get line numbers):
backtrace.h:92 Envoy version: 10604c0031a7ab1b0fdbe1aeced568a6aa41ad4f/1.26.1-dev/Modified/DEBUG/BoringSSL
backtrace.h:96 #0: Envoy::SignalAction::sigHandler() [0x556624937288]
backtrace.h:96 #1: __restore_rt [0x7f9f80bdde30]

ConnectionImpl 0x466bf893450, connecting_: 0, bind_error_: 0, state(): Open, read_buffer_limit_: 1048576
socket_: 
  ListenSocketImpl 0x466bf7ab780, transport_protocol_: raw_buffer
  connection_info_provider_: 

```

Also marked inline - the first ``onAboveWriteBufferHighWatermark`` is called at ``codec_impl.cc:586``, then called at ``connection_impl.cc:548``, and both are calling back ``HttpUpstream::onAboveWriteBufferHighWatermark`` at ``tcp_proxy/upstream.cc:119``.

The following trace logs are provided for the when the assert calls are commented out:

```
... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

tcp_proxy.cc:673             [C2] downstream connection received 119289 bytes, end_stream=false 
codec_impl.cc:586            [C3] send buffer over limit 
upstream.cc:119              HttpUpstream::onAboveWriteBufferHighWatermark
tcp_proxy.cc:338             [C2] Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=0
connection_impl.cc:483       [C3] writing 16393 bytes, end_stream false
codec_impl.cc:1177           [C3] sent frame type=0, stream_id=1, length=16384
codec_impl.cc:593            [C3] send buffer under limit 
upstream.cc:124              HttpUpstream::onBelowWriteBufferLowWatermark
tcp_proxy.cc:349             [C2] Filter::UpstreamCallbacks::onBelowWriteBufferLowWatermark

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

connection_impl.cc:530       [C2] onAboveReadBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=1048585
tcp_proxy.cc:673             [C2] downstream connection received 1048585 bytes, end_stream=false
connection_impl.cc:523       [C2] onBelowReadBufferLowWatermark
connection_impl.cc:362       [C2] readDisable: disable=false disable_count=1 state=0 buffer_length=0
codec_impl.cc:586            [C3] send buffer over limit 
upstream.cc:119              HttpUpstream::onAboveWriteBufferHighWatermark
tcp_proxy.cc:338             [C2] Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=0

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

connection_impl.cc:548       [C3] onAboveWriteBufferHighWatermark    ### This call would cause assert hit, if not commented
upstream.cc:119              HttpUpstream::onAboveWriteBufferHighWatermark
tcp_proxy.cc:338             [C2] Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

codec_impl.cc:593            [C3] send buffer under limit 
upstream.cc:124              HttpUpstream::onBelowWriteBufferLowWatermark
tcp_proxy.cc:349             [C2] Filter::UpstreamCallbacks::onBelowWriteBufferLowWatermark
connection_impl.cc:362       [C2] readDisable: disable=false disable_count=2 state=0 buffer_length=0

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

connection_impl.cc:537       [C3] onBelowWriteBufferLowWatermark
upstream.cc:124              HttpUpstream::onBelowWriteBufferLowWatermark
tcp_proxy.cc:349             [C2] Filter::UpstreamCallbacks::onBelowWriteBufferLowWatermark
connection_impl.cc:362       [C2] readDisable: disable=false disable_count=1 state=0 buffer_length=0

... Removed lines about socket ready/write/ssl etc.. (too many to include here) ...

connection_impl.cc:530       [C2] onAboveReadBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=1048585
tcp_proxy.cc:673             [C2] downstream connection received 1048585 bytes, end_stream=false
connection_impl.cc:523       [C2] onBelowReadBufferLowWatermark
connection_impl.cc:362       [C2] readDisable: disable=false disable_count=1 state=0 buffer_length=0
codec_impl.cc:586            [C3] send buffer over limit 
upstream.cc:119              HttpUpstream::onAboveWriteBufferHighWatermark
tcp_proxy.cc:338             [C2] Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark
connection_impl.cc:362       [C2] readDisable: disable=true disable_count=0 state=0 buffer_length=0
```

In this case there are alternating ``connection_impl.cc:548 - onAboveWriteBufferHighWatermark`` and ``codec_impl.cc:586 - send buffer over limit`` but each one also gets the ``lowWatermark`` call too.

I'll appreciate help verifying if this is the case, before opening a PR to fix/remove the ASSERT calls. cc @alyssawilk, @KBaichoo, @ggreenway

### PR DESCRIPTION
**Pull request by ohadvano**
Additional Description: Fixes #27516. For TCP tunneling scenario it's possible that ``readDisable()`` will be called multiple times, either from codec layer and connection layer. This change only asserts in case TCP tunneling is not enabled, and also increases counters only if read was actually disabled/enabled.
Risk Level: low
Testing: Unit tests
Docs Changes: changelog
Release Notes: none
Platform Specific Features: none

### CODE REVIEW

**Comment by ohadvano** [comment]
/assign @KBaichoo
@KBaichoo Could you please have a first look before I'm proceeding to implementing tests, etc? Thanks

File: source/common/tcp_proxy/tcp_proxy.cc [hunk 1]
```diff
@@ -343,7 +350,11 @@ void Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark() {
 }
 
 void Filter::UpstreamCallbacks::onBelowWriteBufferLowWatermark() {
-  ASSERT(on_high_watermark_called_);
+  // TCP Tunneling may call on high watermark multiple times.
+  if (!parent_->config_->tunnelingConfigHelper()) {
+    ASSERT(on_high_watermark_called_);
```
Comment (KBaichoo) [inline source/common/tcp_proxy/tcp_proxy.cc:355]: Could consider putting the logic of the if into the ASSERT.

e.g. `ASSERT(parent_->config-_->tunnelingConfigHelper() || (!parent_->config_->tunnelingConfigHelper() && on_high_watermark_called_))` which would ensure the entire check is compiled out in optimized mode (as currently done)

**Review by KBaichoo — COMMENTED** [review]
Nice work! Definitely on track.

File: envoy/network/connection.h [hunk 2]
```diff
@@ -84,6 +84,12 @@ class Connection : public Event::DeferredDeletable,
 public:
   enum class State { Open, Closing, Closed };
 
+  enum class ReadDisableStatus {
+    NoTransition,
```
[the remark this answers is not in the document]
Comment (ohadvano) [reply envoy/network/connection.h:88]: got it, I'm not sure ``StillReadEnabled`` makes sense though - the read counter is unsigned int and always positive, so if you call ``readDisable(false)`` when the counter equals 0 that's a bug, so AFAIU you either don't call ``readDisable(false)`` if it's already enabled, or you call ``readDisable(true)`` which would be either transitioning to disable or ``StillReadDisable``. wdyt?

File: source/common/tcp_proxy/tcp_proxy.cc (hunk 1 above)
Comment (ohadvano) [reply source/common/tcp_proxy/tcp_proxy.cc]: done

File: envoy/network/connection.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ohadvano) [reply envoy/network/connection.h:88]: I guess that your suggestion could also work, but can you please elaborate the concern about the current approach?

**Comment by ohadvano** [comment]
/unassign @lizan 
currently out

**Comment by ohadvano** [comment]
@KBaichoo I added tests, and I think it's ready for review
Could you please do another pass and also assign senior which is not Lizan, as he's currently out? Tried unassigning and re-auto assign, but it chose him again

File: envoy/network/connection.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (soulxu) [reply envoy/network/connection.h:88]: I don't have a strong reason for those two options. I think both are ok. Just looking at the discussion about those enum definitions, then this idea is coming up, just utilized the existing interface and return a simple bool here. So just trying to provide one more option if the discussion is going hard here. Not a comment for a required change :)

File: source/common/tcp_proxy/tcp_proxy.cc [hunk 3]
```diff
@@ -293,11 +293,14 @@ void Filter::readDisableDownstream(bool disable) {
     // despite the downstream connection being closed.
     return;
   }
-  read_callbacks_->connection().readDisable(disable);
 
-  if (disable) {
+  Network::Connection::ReadDisableStatus read_disable_status =
```
Comment (KBaichoo) [inline source/common/tcp_proxy/tcp_proxy.cc:297]: const

**Review by KBaichoo — COMMENTED** [review]
Can you add an integration test to reproduce the problem that motivated this work so we can have a regression test? https://github.com/envoyproxy/envoy/issues/27516

Otherwise looks great!

/assign @RyanTheOptimist 
/wait for integration test

File: envoy/network/connection.h [hunk 4]
```diff
@@ -186,7 +193,7 @@ class Connection : public Event::DeferredDeletable,
    * readDisable(false);  // Notes the connection is blocked by one source
    * readDisable(false);  // Marks the connection as unblocked, so resumes reading.
    */
-  virtual void readDisable(bool disable) PURE;
+  virtual ReadDisableStatus readDisable(bool disable) PURE;
```
Comment (RyanTheOptimist) [inline envoy/network/connection.h:196]: Nit: please document the `@return` value

**Review by RyanTheOptimist — COMMENTED** [review]
Agreed about the integration test. Otherwise, one nit.

File: envoy/network/connection.h (hunk 4 above)
Comment (ohadvano) [reply envoy/network/connection.h:197]: done

**Comment by ohadvano** [comment]
@KBaichoo, @RyanTheOptimist I addressed your comments. Regarding the integration tests - I couldn't figure out how to simulate an integration test that hits those callbacks twice. It seems like ``FakeUpstream`` watermark callbacks are null-op. If you have an idea or can help with this, I'll be happy to follow your guidance, otherwise - maybe we should add a backlog to add more comprehensive watermark tests?

**Comment by KBaichoo** [comment]
I think you could achieve this by following something similar to tests in `test/integration/buffer_accounting_integration_test.cc` by using the  socket swap.

/wait

**Comment by ohadvano** [comment]
Thanks for the reference, however it still seems to me that even if I use this approach, the ``onAboveWriteBufferHighWatermark`` will not propagate to the ``tcp_proxy`` filter's callback, because of the null-op here: 
https://github.com/envoyproxy/envoy/blob/1ef7992acc2259b208b1c8c5ea84dcd33f2b072d/test/integration/fake_upstream.h#L238

Unless I misunderstood something, it looks like it would require some refactoring to get this done, could you please share your insight or extra details about what you had in mind for testing this?

**Comment by KBaichoo** [comment]
See: https://github.com/envoyproxy/envoy/blob/1ef7992acc2259b208b1c8c5ea84dcd33f2b072d/test/integration/buffer_accounting_integration_test.cc#L650 as an example.

I think the original problem was TCP proxy <---> HTTP2 upstream cluster; we could size the HTTP2 config / upstream connection buffer to be < downstream connection buffer and use the socket swap to make writes to upstream fail.

We should then be able to get disable from the upstream connection and upstream codec.

**Comment by ohadvano** [comment]
I do think it's possible to get the upstream request to disable and call ``onAboveWriteBufferHighWatermark``, but how's it going to propagate to the ``tcp_proxy::onAboveWriteBufferHighWatermark``? Because it' currently null-op, as I see it the fake upsream is unaware of the callbacks defined by the ``tcp_proxy`` side, and I'm not sure if it's currently possible to pass those callbacks

**Comment by KBaichoo** [comment]
> as I see it the fake upstream is unaware of the callbacks defined by the tcp_proxy side, and I'm not sure if it's currently possible to pass those callbacks

Recommend running the tests I've linked above and you should see the writes don't reach the upstream (they end up being stuck in Envoy due to socket swap injecting errors) which should trigger the watermarking callbacks.

**Comment by ohadvano** [comment]
Ok thanks I'll try this. Just to make sure I understand you, in ``tcp_proxy`` there are two callbacks sets, downstream:
https://github.com/envoyproxy/envoy/blob/1ef7992acc2259b208b1c8c5ea84dcd33f2b072d/source/common/tcp_proxy/tcp_proxy.cc#L309

And upstream:
https://github.com/envoyproxy/envoy/blob/1ef7992acc2259b208b1c8c5ea84dcd33f2b072d/source/common/tcp_proxy/tcp_proxy.cc#L335

Are you saying that this approach should trigger the ``Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark()``?

**Comment by KBaichoo** [comment]
Yes I think we should be able to trigger the `Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark` as done in your production case by:
1) Having the upstream network connection buffer reach watermark
2) Having the HTTP2 codec watermark

Thanks for giving this a shot, it can be tricky.

**Comment by ohadvano** [comment]
@KBaichoo, so I did try to get this done with the ``SocketInterfaceSwap``, but after a lot of variations I still couldn't get that callback to fire twice.
With the following version for example, the callback isn't called at all.

```
TEST_P(TcpTunnelingIntegrationWatermarkTest, testemp) {
  initialize();
  setUpConnection(fake_upstream_connection_);
  write_matcher_->setDestinationPort(fake_upstreams_[0]->localAddress()->ip()->port());
  write_matcher_->setWriteReturnsEgain();

  ASSERT_TRUE(tcp_client_->write(std::string(2000, 'a'), false));
  buffer_factory_->waitUntilTotalBufferedExceeds(1000);

  ASSERT_TRUE(fake_upstream_connection_->close());
  tcp_client_->close();
}
```

With the following variation, the upstream isn't even set up, because response headers are not set (need to call setUpConnection first that also encodes back the response ``CONNECT`` headers):

```
TEST_P(TcpTunnelingIntegrationWatermarkTest, testemp) {
  initialize();
  write_matcher_->setDestinationPort(fake_upstreams_[0]->localAddress()->ip()->port());
  write_matcher_->setWriteReturnsEgain();
  setUpConnection(fake_upstream_connection_);

  ASSERT_TRUE(tcp_client_->write(std::string(2000, 'a'), false));
  buffer_factory_->waitUntilTotalBufferedExceeds(1000);

  ASSERT_TRUE(fake_upstream_connection_->close());
  tcp_client_->close();
}
```

The closest I could get was with this variation, where ``setBufferLimits`` did the trick and actually didn't require the ``SocketInterfaceSwap``.
However, with this version the callback was called only once, when needed twice.

```
TEST_P(TcpTunnelingIntegrationWatermarkTest, testemp) {
  config_helper_.setBufferLimits(10, 10000);
  initialize();
  setUpConnection(fake_upstream_connection_);

  ASSERT_TRUE(tcp_client_->write(std::string(1000, 'a'), false));

  ASSERT_TRUE(fake_upstream_connection_->close());
  tcp_client_->close();
}
```

``TcpTunnelingIntegrationWatermarkTest`` is a test class that is the same as ``Http2OverloadManagerIntegrationTest`` only it also has the ``tcp_proxy`` listener with TCP Tunneling enabled

Any idea what else can be done? Right now I'm feeling a bit stuck on this.

**Comment by ohadvano** [comment]
ping @KBaichoo

**Review by KBaichoo — COMMENTED** [review]
Thank you for working on this. I think you're really close to cracking the reproduction!

If we go with your third attempt: I think all you might need to do is to configure [typed_extension_protocol_options](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto) to specify HTTP2 [initial stream window size](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-msg-config-core-v3-http2protocoloptions) to 65535 and bump up the downstream window to something larger like 128k. This is where the socket swap would be handy to ensure you are able to reach the watermark for the upstream protocol reliably.

**Comment by ohadvano** [comment]
Thanks for guiding this @KBaichoo. I added an integration test to validate this case, and validated that with the previous ``ASSERT`` call the tests are failing due to that assert call hit.

File: test/integration/buffer_accounting_integration_test.cc [hunk 5]
```diff
@@ -606,6 +607,69 @@ TEST_P(ProtocolsBufferWatermarksTest, ResettingStreamUnregistersAccount) {
...
+
+  tcp_client_ = makeTcpConnection(lookupPort("tcp_proxy"));
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+
+  write_matcher_->setWriteReturnsEgain();
+  ASSERT_TRUE(tcp_client_->write(std::string(524288, 'a'), false));
+  tcp_client_->close();
```
Comment (KBaichoo) [inline test/integration/buffer_accounting_integration_test.cc:670]: I think we should wait here on the stat being incremented that there was a single read disable stat increment to the proxy.

File: test/integration/buffer_accounting_integration_test.cc [hunk 6]
```diff
@@ -606,6 +607,69 @@ TEST_P(ProtocolsBufferWatermarksTest, ResettingStreamUnregistersAccount) {
...
+};
+
+INSTANTIATE_TEST_SUITE_P(
+    IpVersions, TcpTunnelingWatermarkIntegrationTest,
+    testing::Combine(testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams(
+                         {Http::CodecType::HTTP2}, {FakeHttpConnection::Type::HTTP2})),
+                     testing::Bool()),
+    protocolTestParamsAndBoolToString);
+
+TEST_P(TcpTunnelingWatermarkIntegrationTest, MultipleReadDisableCalls) {
```
Comment (KBaichoo) [inline test/integration/buffer_accounting_integration_test.cc:656]: nit: consider naming this `CanHandleMultipleReadDisableCalls` or `DoesNotAssertWhenMultipleReadDisableCalls` or `MultipleReadDisableCallsIncrementsStatsOnce` or similar.

File: test/integration/buffer_accounting_integration_test.cc [hunk 7]
```diff
@@ -606,6 +607,69 @@ TEST_P(ProtocolsBufferWatermarksTest, ResettingStreamUnregistersAccount) {
...
+          ConfigHelper::HttpProtocolOptions protocol_options;
+          auto* options =
+              protocol_options.mutable_explicit_http_config()->mutable_http2_protocol_options();
+          options->mutable_initial_stream_window_size()->set_value(65536);
+          ConfigHelper::setProtocolOptions(
+              *bootstrap.mutable_static_resources()->mutable_clusters(0), protocol_options);
+        });
+  }
+
+  IntegrationTcpClientPtr tcp_client_;
```
Comment (KBaichoo) [inline test/integration/buffer_accounting_integration_test.cc:646]: member should be protected

**Review by KBaichoo — COMMENTED** [review]
Great work! Thanks for iterating on this. Looks good to me otherwise!

File: test/integration/buffer_accounting_integration_test.cc (hunk 5 above)
Comment (ohadvano) [reply test/integration/buffer_accounting_integration_test.cc:672]: Thanks, makes sense. Added a check to equal 1. I'm not sure if there's a wait to check it doesn't reach 2 unless we make the test duration longer due to timeout, correct me if I'm wrong.

File: test/integration/buffer_accounting_integration_test.cc (hunk 6 above)
Comment (ohadvano) [reply test/integration/buffer_accounting_integration_test.cc]: done

File: test/integration/buffer_accounting_integration_test.cc (hunk 7 above)
Comment (ohadvano) [reply test/integration/buffer_accounting_integration_test.cc:647]: done

File: test/integration/buffer_accounting_integration_test.cc (hunk 5 above)
Comment (KBaichoo) [reply test/integration/buffer_accounting_integration_test.cc:672]: Nope there's no wait for it to not change (and that would just be effectively a sleep wasting CI time), thanks for checking that it breaks if we do.

**Review by KBaichoo — APPROVED** [review]

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by ohadvano** [comment]
It seems like the CI is consecutively failing on ``example_config_test`` although it passed before and I'm not really sure it has anything to do with this PR changes. Any idea what's going on?

**Comment by ohadvano** [comment]
@phlax hopefully you could help here, how can the Verify/examples step retried?

**Pull request merged**